packages/remix-react/components.tsx

```diff
@@ -904,11 +954,203 @@ export function useActionData<T = AppData>(): SerializeFrom<T> | undefined {
  * navigation indicators and optimistic UI on data mutations.
  *
  * @see https://remix.run/api/remix#usetransition
+ * @deprecated Deprecated in favor of useNavigation
  */
 export function useTransition(): Transition {
-  // TODO: For SSR we can just return IDLE_TRANSITION for now, we'll handle
-  // this as part of hydration work
-  return IDLE_TRANSITION;
+  let navigation = useNavigation();
+
+  // Have to avoid useMemo here to avoid introducing unstable transition object
+  // identities in StrictMode, since navigation will be stable but using
+  // [navigation] as the dependency array will _still_ re-run on concurrent
+  // renders, and that will create a new object identify for transition
+  let lastNavigationRef = React.useRef<Navigation>();
+  let lastTransitionRef = React.useRef<Transition>();
+
+  if (lastTransitionRef.current && lastNavigationRef.current === navigation) {
+    return lastTransitionRef.current;
+  }
+
+  lastNavigationRef.current = navigation;
+  lastTransitionRef.current = convertNavigationToTransition(navigation);
+
+  return lastTransitionRef.current;
```

Comment:
Concurrent mode prevents `useMemo` here since it re-runs even if the `navigation` doesn't change and creates a new `transition` object identity 😖

Reply:
Lets convert this back to `useMemo` and update our integration test - but Matt/Jacob to put together a smaller reproduction to discuss separately.